### ISSUE TITLE
ClientLoginError UberSDK

### ISSUE THREAD

**Issue by adeelasim** [uber/rides-ios-sdk#57]
I have followed the exact implementation of Uber iOS SDK in my iOS, and made sure that Callback URI schemes are registered in the uber developer dashboard which i have used in the app. But its throwing UBSDKRidesClientError. Here is my code for the button:

```objective-c
UBSDKRideRequestButton *button = [[UBSDKRideRequestButton alloc] init];
UBSDKRidesClient *ridesClient = [[UBSDKRidesClient alloc] init];

__block UBSDKRideParametersBuilder *builder = [[UBSDKRideParametersBuilder alloc] init];
builder = [builder setPickupLocation: pickupLocation];
builder = [builder setDropoffLocation: dropoffLocation];
[ridesClient fetchCheapestProductWithPickupLocation: pickupLocation completion:^(UBSDKUberProduct* _Nullable product, UBSDKResponse* _Nullable response) {

    if (product) {
        builder = [builder setProductID: product.productID];
        button.rideParameters = [builder build];
        [button loadRideInformation];
    }
}];
```
Will appreciate any help

**Comment by jbrophy17** [comment]
Hi @adeelasim, sorry to hear you're having trouble. Can you let me know what the `code` of your UBSDKRidesClientError is? 

My initial guess is an authorization error, are you signed in while attempting this? Or do you have the `UberServerToken` entry in your Info.plist?

**Comment by adeelasim** [comment]
Hi, Thanks for getting back with me.
I just checked its giving me unauthorized error. I have already signed in on Uber app.
I didn't set the UberServerToken in Info.plist because its not mentioned in documentation of Uber SDK. So I have just set the UberServerToken and its working now but one thing is happening, I am already passing pickup and drop of locations as you can see in my above posted code but Uber App not pre filling the destination. Any idea why?

**Comment by jbrophy17** [comment]
Sorry for the confusion about the Server Token (see #51), glad to hear it's working now. As for the destination not being populated, this is because you need to provide a nickname or formatted address for the destination (see #1). This is missing from the example, however, and needs to be updated. 

I've opened #58

**Comment by adeelasim** [comment]
Thanks for the super fast help John.
It all works now!

**Comment by jbrophy17** [comment]
Happy to help

**Comment by adeelasim** [comment]
Hey, just a quick question. Can i change only text color of the Uber Button to Grey?

**Comment by jbrophy17** [comment]
Hey @adeelasim, happy to help but could you open a new issue? Thanks!

**Comment by hakimny** [comment]
Hi John,

I am running into the same problem with SDK 5.0.1 , server response:
response : ({
    code = unauthorized;
    message = "No authentication provided.";
})

I made sure to add the UberServerToken , I did test that my token is working properly using POSTMAN using this request : https://api.uber.com/v1/estimates/time?server_token=MY_SERVER_TOKEN&start_latitude=45.49214145&start_longitude=-73.58121160
but when I use it 
// Swift
let button = RideRequestButton()
let ridesClient = RidesClient()
let pickupLocation = CLLocation(latitude: 37.787654, longitude: -122.402760)
let dropoffLocation = CLLocation(latitude: 37.775200, longitude: -122.417587)
let builder = RideParametersBuilder().setPickupLocation(pickupLocation).setDropoffLocation(dropoffLocation)
ridesClient.fetchCheapestProduct(pickupLocation: pickupLocation, completion: {
    product, response in
    if let productID = product?.productID {
        builder = builder.setProductID(productID)
        button.rideParameters = builder.build()
        button.loadRideInformation()
}
})

The funny thing is that it used to work fine and the RideRequestButton display estimated time & fares then starting a week or so it stopped working giving the error mention above.

As I said after reading this post I updated my plist to include UberServerToken, but no luck in making it work so far.

Thank you so much for your prompt answer, I am about to upload my Prod app to app store and it is the only thing missing.

Hakim

**Comment by jbrophy17** [comment]
Hi Hakim,
Sorry to hear you're having trouble, looks like it might be some sort of
access token issue since it was working before. Have you tried signing out
and signing back in?

**Comment by hakimny** [comment]
Hi John,
Yes I did and no dice still not working, I am clueless , the RideRequestWidget because the authentication failed come out blank.

Anything else you can think of?

Thx I appreciate your prompt help
Hakim

**Comment by hakimny** [comment]
![img_1396](https://cloud.githubusercontent.com/assets/242229/17291812/14c916e0-57b4-11e6-8200-e327df46700a.PNG)
This is what I get if I tap uber rides button, I deleted my app many time just to clear anything saved on cache or NSDefaults() , but no dice. I have to point out that documentation needs to be updated, as it is getting a bit confusing.

thx for all your help

**Comment by jbrophy17** [comment]
When you open the ride request widget does it prompt you for authentication? or does it just always stay blank like that? 

Can you also open a separate issue for the confusion you are seeing in the documentation? Thanks

**Comment by hakimny** [comment]
Hi John,

It stays blank like that and does not prompt for authentication.

PS: Will open a separate issue for the documentation

thx

**Comment by jbrophy17** [comment]
That is strange, if it's not prompting for authentication, a bad access token must be cached somewhere. I know you said you tried deleting your app to clear everything, but could you also try explicitly calling `TokenManager.deleteToken()`?
Unfortunately, I'm not able to reproduce what you are seeing on my end

**Comment by hakimny** [comment]
Hi John,
calling TokenManager.deleteToken()  did the trick , thank you so much for your help, a couple of things I learned from all this:
UberServerToken needs to be added to plist
RideRequestViewRequestingBehavior should be called with LoginManger
if only the Ride Request Widget is needed then LoginManager should be called with login Type : .Implicit as such :

let loginManager = LoginManager(loginType: .Implicit)
let requestBehavior = RideRequestViewRequestingBehavior(presentingViewController: self, loginManager: loginManager)
rideRequestButton.requestBehavior = requestBehavior

**Comment by hakimny** [comment]
![img_1397](https://cloud.githubusercontent.com/assets/242229/17311767/c99fc19c-581b-11e6-928a-bb720c91896d.PNG)
![img_1398](https://cloud.githubusercontent.com/assets/242229/17311768/c9a46936-581b-11e6-831a-8178295c4bb2.PNG)
finally working the way it supposed to, 

thx john for all your help

Hakim

**Comment by jbrophy17** [comment]
Glad to hear things are working. The latest release should prevent the widget from getting into that weird state by automatically refreshing invalid tokens (if you have a refresh token)

Fixed in [0.5.2](https://github.com/uber/rides-ios-sdk/releases/tag/v0.5.2)
